### PR TITLE
Group Makefile targets

### PR DESCRIPTION
**Pull request by Skarlso**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

A nicer Makefile help output:

```
➜  cluster-api-provider-aws git:(group_makefile) ✗ make

Usage:
  make <target>

Testing
  test                                      Run tests
  generate-test-flavors                     Generate test template flavors
  test-conformance                          Run clusterctl based conformance test on workload cluster (requires Docker).
  test-cover                                Run tests with code coverage and code generate  reports

Binaries
  binaries                                  Builds and installs all binaries
  manager-aws-infrastructure                Build manager binary.
  clusterawsadm                             Build clusterawsadm binary.

Linting
  lint                                      Lint codebase
  lint-fix                                  Lint the codebase and run auto-fixers if supported by the linter

Generate
  modules                                   Runs go mod to ensure proper vendoring.
  generate                                  Generate code
  generate-go-apis                          Runs Go related generate targets

Docker
  docker-build                              Build the docker image for controller-manager
  docker-push                               Push the docker image
  docker-push-core-manifest                 Push the fat manifest docker image.

Release
  list-staging-releases                     List staging images for image promotion
  release                                   Builds and push container images using the latest git tag for the commit.
  release-changelog                         Builds the changelog for a release
  release-binaries                          Builds the binaries to publish with a release
  release-staging                           Builds and push container images and manifests to the staging bucket.
  release-staging-nightly                   Tags and push container images to the staging bucket.
  upload-staging-artifacts                  Upload release artifacts to the staging bucket
  create-gh-release                         Create release on Github
  upload-gh-artifacts                       Upload artifacts to Github release
  release-templates                         Generate release templates

Cleanup / Verification
  clean                                     Remove all generated files
  clean-bin                                 Remove all generated binaries
  clean-temporary                           Remove all temporary files and folders
  serve-book                                Run a server with the documentation book
  clean-release                             Remove the release folder
  verify-conversions                        Verifies expected API conversion are in place
  verify-shellcheck                         Verify shell files
  clean-artifacts                           Remove the _artifacts directory

Help
  help                                      Display this help
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/3188

**Special notes for your reviewer**:

**Checklist**:

- [x] squashed commits
- [x] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:

```release-note
Adding a nicer Makefile help output so it's easy to locate a certain target.
```

### CODE REVIEW

**Comment by Skarlso** [comment]
@shivi28  👋  :) 

These are the categories in the Makefile itself. I wanted to keep that as it is but I'm happy to reorder as needed. :)

**Comment by sedefsavas** [comment]
It would be nice to make the categories look more like cluster-api, feel free to reorder as needed.

```make
Usage:
  make <target>

Targets:

generate:
  generate                                       Run all generate-manifests-*, generate-go-deepcopy-* and generate-go-conversions-* targets
  generate-manifests                             Run all generate-manifests-* targets
  generate-manifests-core                        Generate manifests e.g. CRD, RBAC etc. for core
  generate-manifests-kubeadm-bootstrap           Generate manifests e.g. CRD, RBAC etc. for kubeadm bootstrap
  generate-manifests-kubeadm-control-plane       Generate manifests e.g. CRD, RBAC etc. for kubeadm control plane
  generate-go-deepcopy                           Run all generate-go-deepcopy-* targets
  generate-go-deepcopy-core                      Generate deepcopy go code for core
  generate-go-deepcopy-kubeadm-bootstrap         Generate deepcopy go code for kubeadm bootstrap
  generate-go-deepcopy-kubeadm-control-plane     Generate deepcopy go code for kubeadm control plane
  generate-go-conversions                        Run all generate-go-conversions-* targets
  generate-go-conversions-core                   Generate conversions go code for core
  generate-go-conversions-kubeadm-bootstrap      Generate conversions go code for kubeadm bootstrap
  generate-go-conversions-kubeadm-control-plane  Generate conversions go code for kubeadm control plane
  diagrams                                       Generate diagrams for *.plantuml files

lint and verify:
  modules                                        Run go mod tidy to ensure modules are up to date
  lint                                           Lint the codebase
  lint-fix                                       Lint the codebase and run auto-fixers if supported by the linter
  apidiff                                        Check for API differences
  format-tiltfile                                Format the Tiltfile
  verify                                         Run all verify-* targets
  verify-modules                                 Verify go modules are up to date
  verify-gen                                     Verfiy go generated files are up to date
  verify-conversions                             Verifies expected API conversion are in place
  verify-boilerplate                             Verify boilerplate text exists in each file
  verify-shellcheck                              Verify shell files
  verify-tiltfile                                Verify Tiltfile format
  verify-book-links                              Verify book links

build:
  clusterctl                                     Build the clusterctl binary
  managers                                       Run all manager-* targets
  manager-core                                   Build the core manager binary into the ./bin folder
  manager-kubeadm-bootstrap                      Build the kubeadm bootstrap manager binary into the ./bin folder
  manager-kubeadm-control-plane                  Build the kubeadm control plane manager binary into the ./bin folder
  docker-build                                   Run all docker-build-* targets
  docker-build-core                              Build the docker image for core controller manager
  docker-build-kubeadm-bootstrap                 Build the docker image for kubeadm bootstrap controller manager
  docker-build-kubeadm-control-plane             Build the docker image for kubeadm control plane controller manager
  e2e-framework                                  Builds the CAPI e2e framework

test:
  test                                           Run unit and integration tests
  test-e2e                                       Run the e2e tests
  test-junit                                     Run tests with verbose setting and generate a junit report
  test-cover                                     Run tests with code coverage and code generate reports
  test-verbose                                   Run tests with verbose settings
  serve-book                                     Build and serve the book with live-reloading enabled
  kind-cluster                                   Create a new kind cluster designed for testing with Tilt
  docker-build-e2e                               Rebuild all Cluster API provider images to be used in the e2e tests

release:
  release                                        Build and push container images using the latest git tag for the commit
  release-manifests                              Build the manifests to publish with a release
  release-manifests-dev                          Build the development manifests and copies them in the release folder
  release-binaries                               Build the binaries to publish with a release
  release-staging                                Build and push container images to the staging bucket
  release-staging-nightly                        Tag and push container images to the staging bucket. Example image tag: cluster-api-controller:nightly_main_20210121
  release-alias-tag                              Add the release alias tag to the last build tag
  docker-build-all                               Build all the architecture docker images
  docker-push                                    Push the docker images
  docker-push-all                                Push all the architecture docker images
  docker-push-manifest-core                      Push the multiarch manifest for the core docker images
  docker-push-manifest-kubeadm-bootstrap         Push the multiarch manifest for the the kubeadm bootstrap docker images
  docker-push-manifest-kubeadm-control-plane     Push the multiarch manifest for the the kubeadm control plane docker images

clean:
  clean                                          Remove all generated files
  clean-bin                                      Remove all generated binaries
  clean-book                                     Remove all generated GitBook files
  clean-release                                  Remove the release folder
  clean-release-git                              Restores the git files usually modified during a release
  clean-generated-conversions                    Remove files generated by conversion-gen from the mentioned dirs. Example SRC_DIRS="./api/v1alpha4"

hack/tools:
  controller-gen                                 Build a local copy of controller-gen
  conversion-gen                                 Build a local copy of conversion-gen
  conversion-verifier                            Build a local copy of conversion-verifier
  gotestsum                                      Build a local copy of gotestsum
  go-apidiff                                     Build a local copy of apidiff
  golangci-lint                                  Build a local copy of golangci-lint
  envsubst                                       Build a local copy of envsubst
  kustomize                                      Build a local copy of kustomize
  setup-envtest                                  Build a local copy of setup-envtest
  tilt-prepare                                   Build a local copy of tilt-prepare
  kpromo                                         Build a local copy of kpromo

```

**Comment by Skarlso** [comment]
@sedefsavas how about this:

```
➜  cluster-api-provider-aws git:(group_makefile) ✗ make

Usage:
  make <target>

generate
  generate                                  Generate code
  generate-go-apis                          Runs Go related generate targets

list and verify
  verify-conversions                        Verifies expected API conversion are in place
  verify-shellcheck                         Verify shell files

build
  binaries                                  Builds and installs all binaries
  manager-aws-infrastructure                Build manager binary.
  clusterawsadm                             Build clusterawsadm binary.

lint
  lint                                      Lint codebase
  lint-fix                                  Lint the codebase and run auto-fixers if supported by the linter

test
  test                                      Run tests
  generate-test-flavors                     Generate test template flavors
  test-conformance                          Run clusterctl based conformance test on workload cluster (requires Docker).
  test-cover                                Run tests with code coverage and code generate  reports
  serve-book                                Run a server with the documentation book

release
  list-staging-releases                     List staging images for image promotion
  release                                   Builds and push container images using the latest git tag for the commit.
  release-changelog                         Builds the changelog for a release
  release-binaries                          Builds the binaries to publish with a release
  release-staging                           Builds and push container images and manifests to the staging bucket.
  release-staging-nightly                   Tags and push container images to the staging bucket.
  upload-staging-artifacts                  Upload release artifacts to the staging bucket
  create-gh-release                         Create release on Github
  upload-gh-artifacts                       Upload artifacts to Github release
  release-templates                         Generate release templates

clean
  clean                                     Remove all generated files
  clean-bin                                 Remove all generated binaries
  clean-temporary                           Remove all temporary files and folders
  clean-release                             Remove the release folder
  clean-artifacts                           Remove the _artifacts directory

tools
  docker-build                              Build the docker image for controller-manager
  docker-push                               Push the docker image
  docker-push-core-manifest                 Push the fat manifest docker image.
  modules                                   Runs go mod to ensure proper vendoring.

Help
  help                                      Display this help
```

File: Makefile [hunk 1]
```diff
@@ -583,59 +571,58 @@ clean-temporary: ## Remove all temporary files and folders
...
+##@ tools:
 
-.PHONY: verify-modules
-verify-modules: modules
-	@if !(git diff --quiet HEAD -- go.sum go.mod hack/tools/go.mod hack/tools/go.sum); then \
-		git diff; \
-		echo "go module files are out of date"; exit 1; \
-	fi
+.PHONY: docker-build
+docker-build: docker-pull-prerequisites ## Build the docker image for controller-manager
```
Comment (sedefsavas) [inline Makefile:585]: This can go under build category?

File: Makefile [hunk 2]
```diff
@@ -583,59 +571,58 @@ clean-temporary: ## Remove all temporary files and folders
...
+	docker build --build-arg ARCH=$(ARCH) --build-arg LDFLAGS="$(LDFLAGS)" . -t $(CORE_CONTROLLER_IMG)-$(ARCH):$(TAG)
 
-.PHONY: verify-conversions
-verify-conversions: $(CONVERSION_VERIFIER)  ## Verifies expected API conversion are in place
-	echo verification of api conversions initiated
-	$(CONVERSION_VERIFIER)
+.PHONY: docker-push
+docker-push: ## Push the docker image
```
Comment (sedefsavas) [inline Makefile:589]: This can go under release category.

**Comment by Skarlso** [comment]
I added some descriptions which where missing, therefore, they weren't in the help at all. That might have been maybe intentional? I can remove them if desired. Further, `.build` and `.*` targets aren't rendered by the help. I can fix that if desired, if not then ¯\\\_(ツ)\_/¯ 

```
➜  cluster-api-provider-aws git:(group_makefile) ✗ make

Usage:
  make <target>

generate:
  generate                                  Generate code
  generate-go-apis                          Alias for .build/generate-go-apis
  defaulters                                Generate all Go types

lint and verify:
  modules                                   Runs go mod to ensure proper vendoring.
  lint                                      Lint codebase
  lint-fix                                  Lint the codebase and run auto-fixers if supported by the linter
  verify-boilerplate                        Verify boilerplate
  verify-modules                            Verify go modules integrity
  verify-conversions                        Verifies expected API conversion are in place
  verify-shellcheck                         Verify shell files
  verify-gen                                Verify generated files

build:
  binaries                                  Builds and installs all binaries
  managers                                  Alias for manager-aws-infrastructure
  manager-aws-infrastructure                Build manager binary
  clusterawsadm                             Build clusterawsadm binary
  docker-pull-prerequisites                 Build prerequisites using docker
  docker-build                              Build the docker image for controller-manager

test:
  test                                      Run tests
  generate-test-flavors                     Generate test template flavors
  test-conformance                          Run clusterctl based conformance test on workload cluster (requires Docker).
  test-cover                                Run tests with code coverage and code generate  reports
  serve-book                                Run a server with the documentation book

release:
  build-toolchain                           Build the toolchain
  check-github-token                        Check if the github token is set
  check-previous-release-tag                Check if the previous release tag is set
  check-release-tag                         Check if the release tag is set
  create-gh-release                         Create release on Github
  compiled-manifest                         Compile the manifest files
  docker-push                               Push the docker image
  docker-push-core-manifest                 Push the fat manifest docker image.
  docker-push-manifest                      Push the manifest image
  image-patch-source-manifest               Patch the source manifest
  image-patch-kustomization                 Alias for image-patch-kustomization-without-webhook
  image-patch-kustomization-without-webhook  Patch the image in the kustomization file
  image-patch-pull-policy                   Patch the pull policy
  list-staging-releases                     List staging images for image promotion
  list-image                                List images for RELEASE_TAG
  release                                   Builds and push container images using the latest git tag for the commit.
  release-policies                          Release policies
  release-manifests                         Release manifest files
  release-changelog                         Builds the changelog for a release
  release-binaries                          Builds the binaries to publish with a release
  release-binary                            Release binary
  release-staging                           Builds and push container images and manifests to the staging bucket.
  release-staging-nightly                   Tags and push container images to the staging bucket.
  release-templates                         Generate release templates
  upload-staging-artifacts                  Upload release artifacts to the staging bucket
  upload-gh-artifacts                       Upload artifacts to Github release
  staging-manifests                         Staging manifests

clean:
  clean                                     Remove all generated files
  clean-bin                                 Remove all generated binaries
  clean-temporary                           Remove all temporary files and folders
  clean-release                             Remove the release folder
  clean-artifacts                           Remove the _artifacts directory

help:
  help                                      Display this help
```

**Comment by Skarlso** [comment]
I just changed the makefile ordering, I don't think I messed with the tests. Unless, I accidentally deleted a make target maybe?

**Comment by Skarlso** [comment]
Yaaay, seems like the tests passed.

**Comment by shivi28** [comment]
Hey @Skarlso Thanks for the PR.
Yes tests were not failing because of your changes, there was some issue with test pipeline, it's been fixed now.

I have one suggestion, can we sort the categories command alphabetically. Wdyt?
eg 
```
test:
  test                                      Run tests
  generate-test-flavors                     Generate test template flavors
  test-conformance                          Run clusterctl based conformance test on workload cluster (requires Docker).
  test-cover                                Run tests with code coverage and code generate  reports
  serve-book                                Run a server with the documentation book
```
can be re-written as
```
test:
  generate-test-flavors                     Generate test template flavors
  test                                      Run tests
  test-conformance                          Run clusterctl based conformance test on workload cluster (requires Docker).
  test-cover                                Run tests with code coverage and code generate  reports
  serve-book                                Run a server with the documentation book
```
Same we can do with others also.

**Comment by shivi28** [comment]
And once all changes are done, we can squash all the commits to single comment

**Pull request merged**